File: kakao/2018/kakao2018_6.cpp
#include <string>
#include <vector>
using namespace std;

int charcmp(char a,char b)
{
	if(a==b)return 0;
	else if(isupper(a)&&!isupper(b)&&a==(b-32))return 0;
	else if(!isupper(a)&&isupper(b)&&a==(b+32))return 0;
	return 1;
}

int solution(int m,int n,vector<string> board)
{
    int answer = 0;
    bool flag=false;
	do
	{
		flag=0;
		for(int i=0;i<m-1;i++)
		{
			for(int j=0;j<n-1;j++)
			{
				if(board[i][j]!=' '&&!charcmp(board[i][j],board[i][j+1])&&!charcmp(board[i][j],board[i+1][j])&&!charcmp(board[i+1][j+1],board[i][j]))
				{
					if(isupper(board[i][j]))board[i][j]+=32,answer++;
					if(isupper(board[i][j+1]))board[i][j+1]+=32,answer++;
					if(isupper(board[i+1][j]))board[i+1][j]+=32,answer++;
					if(isupper(board[i+1][j+1]))board[i+1][j+1]+=32,answer++;
					flag=true;
				}
			}
		}
		if(flag)
		{
			for(int k=0;i<m-1;i++)
			{
				for(int i=0;i<m-1;i++)
				{
					for(int j=0;j<n;j++)
					{
						if(board[i+1][j]==' '||!isupper(board[i+1][j]))
						{
							board[i+1][j]=board[i][j];
							board[i][j]=' ';
						}
					}
				}
			}
		}
	}while(flag);
    return answer;
}

File: kakao/2018/kakao2018_6.py
def solution(m, n, board):
    x=[[*i] for i in board];flag=1
    while flag:
        a=[];flag=0
        for i in range(m-1):
            for j in range(n-1):
                if x[i][j]==x[i][j+1]==x[i+1][j]==x[i+1][j+1]!=1:a.append([i,j]);flag=1
        for k in a:
            i,j=k[0],k[1]
            x[i][j],x[i][j+1],x[i+1][j],x[i+1][j+1]=1,1,1,1
        for j in range(n):
            for k in range(m):
                for i in range(m-1):
                    if x[i+1][j]==1:x[i+1][j],x[i][j]=x[i][j],1
    return sum(x,[]).count(1)